### ISSUE TITLE
Vue support

### ISSUE THREAD

**Issue by andreasvirkus** [tuchk4/storybook-readme#37]
Just wondering if there's any near-future plans to support Vue?

Thanks!

**Comment by andreasvirkus** [comment]
Would you also consider supporting from custom blocks (like `documentation`)?
`vue-loader` has the extraction support already built in.

https://vue-loader.vuejs.org/en/configurations/custom-blocks.html

**Comment by ndelangen** [comment]
@andreasvirkus Is this something you could help @tuchk4 with?

**Comment by andreasvirkus** [comment]
My keyboard is yours to command, @tuchk4!

**Comment by tuchk4** [comment]
@andreasvirkus  Any changes are up to you.

I will be happy if you provide PR with `example/vue` and then we will decide what problems we have and what features we want.

**Comment by andreasvirkus** [comment]
@tuchk4 Should I first wait for the Vue support to be released? Or can I base my branch off of a feature branch where the Vue support exists? Couldn't find a relevant branch under this repo.

**Comment by andreasvirkus** [comment]
@tuchk4 ping. Any news on Vue support? Also would like to go over the code with you quickly if you'd be willing to guide me to making the `vue-loader` and custom `<docs>` tag work.

**Comment by tuchk4** [comment]
Sorry for late response. I am going to implement vue support this week.
I push new branch `vue-support`

But I have few issues (because I have not experience with vue):

### 1. Failed to mount component
@andreasvirkus I will be happy if you will help me with it. 
Then we should decide how to include `<docs>`.

```
vue.esm.js:571 [Vue warn]: Failed to mount component: template or render function not defined.

found in

---> <MyButton> at example-vue/components/MyButton/MyButton.vue
       <Anonymous>
         <Root>
```

### 2. Storybook bin files
@ndelangen This question to you :)
Packages `@storybook/react` and `@storybook/vue` have binaries with same names.
- https://github.com/storybooks/storybook/blob/master/app/vue/package.json#L13
- https://github.com/storybooks/storybook/blob/master/app/react/package.json#L13

Is it possible to install both of them? I need to run react and vue storybooks for examples and testing readme addon.

**Comment by tuchk4** [comment]
Update:

**1. Failed to mount component**
ref: https://github.com/vuejs/vue-loader/issues/844#issuecomment-308976557

This error was because there were two rules for `vue-loader`  at webpack config (one rule - default storybook webaock config, second rule - at my *.storybook/webpack.config.js*)

##  \<docs\> section

```vue
<template>
  <div><slot></slot></div>
</template>

<docs>
  ## This is docs inside component
</docs>

<script>
  export default {};
</script>
```

It is easy to implement with [runtime-available-docs](https://vue-loader.vuejs.org/en/configurations/custom-blocks.html#runtime-available-docs):

```js
{
  test: /\.vue$/,
  loader: 'vue-loader',
  options: {
    loaders: {
      docs: docsLoader,
    },
  },
}
```

To add `docsLoader` - I should use [Full Control Mode for Storybook webpack configuration](https://storybook.js.org/configurations/custom-webpack-config/#full-control-mode) and change vue loader options (It is not possible to use [Extend Mode](https://storybook.js.org/configurations/custom-webpack-config/#extend-mode) becase of *vue-loader* rule duplication). 
This is not cool because developers should add this to storybook weback config at every vue storybook with readme addon.

```js
module.exports = (storybookBaseConfig, configType) => {
  storybookBaseConfig.module.rules = storybookBaseConfig.module.rules.map(rule => {
    if (rule.loader.indexOf('vue-loader') !== -1) {
      return {
        ...rule,
        options : {
          ...rule.options
          loaders: {
            docs: docsLoader,
          },
        }
      }
    }

    return rule;
  });

  return storybookBaseConfig;
};
```

@ndelangen any ideas how to configure it in most friendly way? 

---

To add component's docs via *withDocs* or *withReadme*:

```js
import MyButton from '../components/MyButton/MyButton.vue';

storiesOf('withDocs', module)
  .addDecorator(withDocs(MyButton.__docs))
  .add('Button', () => {
    return {
      components: { MyButton },
      template: `<my-button>My Button</my-button>`,
    };
  });

```

**Comment by tuchk4** [comment]
Another problem I faced:

**Addons at right (bottom) panel are implemented with React.** Does not matter if storybook for vue or react.
It is easy to implemeted *withReadme* api.

Problems with **withDocs** api.
**withDocs** api wraps story function:

```
<div>
  {/* Mardown before Component Story */}
  <ReadmeContainer markdown={readmeBeforePreview} />
  
  {/* Story */}
  <PreviewComponent>{storyFn(context)}</PreviewComponent>

  {/* Mardown after Component Story */}
  <ReadmeContainer markdown={readmeAfterPreview)} />

  {/* Common Footer */}
  {config.footer &&
  <FooterComponent>
    <ReadmeContainer markdown={common.footer} />
  </FooterComponent>}
</div>
```

**\<ReadmeContainer markdown\/\>** - main component that renders markdown as html with code highlighting.

Should I re-implement all components (at least \<ReadmeContainer markdown\/\>) with vue? :(
Or can I render react component as vue component? :)

**Comment by tuchk4** [comment]
Update

I have solutions ;)

Will commit in few hours

**Comment by tuchk4** [comment]
Update

At master is supports both react and vue.

```
yarn storybook-react
yarn storybook-vue
```

I will be happy if somebody review code related to vue part (it was my first experience with vue 🙃):
- [packages/example-vue](https://github.com/tuchk4/storybook-readme/tree/master/packages/example-vue)
- [packages/storybook-readme/src/env/vue](https://github.com/tuchk4/storybook-readme/tree/master/packages/storybook-readme/src/env/vue)

vue support comes with huge update:
- now use lerna (private packages for react-example, vue-example, and public package for storybook-readme)
- use webpack loaders to transform markdown to html. Also this is related to https://github.com/storybooks/storybook/pull/2299
- refactor old code

The only one problem (except bugs) - webpack configs:
- [for vue](https://github.com/tuchk4/storybook-readme/blob/master/packages/example-vue/.storybook/webpack.config.js)
- [for webpack](https://github.com/tuchk4/storybook-readme/blob/master/packages/example-react/.storybook/webpack.config.js)

Looking for help to simplify them.

**Comment by ndelangen** [comment]
@tuchk4 Some inspiration:
https://storybooks.slack.com/archives/C5TE0A4AW/p1510616980000298

Maybe you should as in the [Vue channel in slack](https://storybooks.slack.com/messages/C5TE0A4AW)

**Comment by chinchang** [comment]
@tuchk4 I tried the vue support, but instead of seeing the docs in html, I see the docs as html:

![screenshot 2017-12-08 11 26 07](https://user-images.githubusercontent.com/379918/33752966-d8d3dde6-dc0a-11e7-9499-9ae3254fbf57.png)

(Using your examples as base config)

What might be missing?

**Comment by tuchk4** [comment]
@chinchang did you try version from master? (because master is not released yet)
with this config https://github.com/tuchk4/storybook-readme/tree/master/packages/example-vue/.storybook?

**Comment by chinchang** [comment]
Yes, I used master. What branch should I try?

**Comment by tuchk4** [comment]
Vue support is released with `3.2.0`

here is how configure webpack https://github.com/tuchk4/storybook-readme#webpack-configuration-for-vue-storybook

Close this issue. Please reopen or create new if any issues.
